### PR TITLE
Role export access in Leads, Forms, Reports

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:
Added new permission for export in form results. This permission give ability enable/disable export permission for Contacts, Forms, Reports.

image

#### Steps to test this PR:
1.  Create new user with role and enable all permission for Contacts, Forms, Reports with checked  new Export access: Enable
2.  Login as new user and see If exports buttons exists and exports works properly in all sections: Contacts (batch export), Form (results exports) and Reports
3. Open HTML exports of Form results and Reports.
4. Edit in incognito user permission and disable export 
5. See If the export button are hidden in Form results and  Reports and Contacts lists in batch.
6. See If opened HTML export from 3. Return access denied after refresh

### CODE REVIEW

**Review by PatrickJenkner — APPROVED** [review]
Worked exactly as described when tested 👍

**Comment by RCheesley** [comment]
Just came to test this but we've got some conflicts to resolve, and it will need rebasing to 5.1. Looks like we just need 1 more test to get it mergeable though, so happy to do that once it's ready.

**Comment by kuzmany** [comment]
@RCheesley rebased

**Comment by RCheesley** [comment]
Thanks @kuzmany I've added it back to the OSF board and set the status accordingly! 🚀

File: app/bundles/CoreBundle/Security/Permissions/AbstractPermissions.php [hunk 1]
```diff
@@ -302,6 +302,32 @@ public function parseForJavascript(array &$perms): void
     {
     }
 
+    /**
+     * @param string            $level
+     * @param array<int|string> $permissions
+     */
+    protected function addCustomPermission($level, $permissions): void
```
Comment (tomekkowalczyk) [inline app/bundles/CoreBundle/Security/Permissions/AbstractPermissions.php]: We don't need annotations $level and we can parameterize the function arguments

File: app/bundles/CoreBundle/Twig/Helper/SecurityHelper.php [hunk 2]
```diff
@@ -28,6 +28,16 @@ public function getName(): string
         return 'security';
     }
 
+    /**
+     * Helper function to check if user is an Admin.
+     *
+     * @return bool
+     */
+    public function isAdmin()
```
Comment (tomekkowalczyk) [inline app/bundles/CoreBundle/Twig/Helper/SecurityHelper.php]: We don't need the return annotation and specify the return type

**Review by tomekkowalczyk — COMMENTED** [review]
@kuzmany good job I left two comments for you to check out

**Review by nick-vanpraet — CHANGES_REQUESTED** [review]
Tested on gitpod, permissions work for forms and reports but not for contacts. There the export option is always visible, regardless of export permission.

**Comment by RCheesley** [comment]
@kuzmany several bits of feedback here to review! :)

File: app/bundles/CoreBundle/Twig/Helper/SecurityHelper.php (hunk 2 above)
Comment (kuzmany) [reply app/bundles/CoreBundle/Twig/Helper/SecurityHelper.php]: I am unsure If I can add a strict return type for the public method. @escopecz Should I?

**Comment by kuzmany** [comment]
@nick-vanpraetThanks! I have fixed it in the latest commit. Could you please retest it?

**Comment by nick-vanpraet** [comment]
On the contact detail view page the export button is still always available. It's gone in other places if you don't have the correct permission.

Screenshot 2024-02-02 at 11 35 07

Forms and reports still works correctly.

File: app/bundles/CoreBundle/Twig/Helper/SecurityHelper.php (hunk 2 above)
Comment (escopecz) [reply app/bundles/CoreBundle/Twig/Helper/SecurityHelper.php]: If it's a new method, then yes. Or is there some reason why you thing it could break BC?

**Comment by escopecz** [comment]
@kuzmany please, resolve the conflict

**Comment by escopecz** [comment]
I'm testing the migration and it didn't enable the export permission for me. How is it suppose to work? I don't see any condition. It should enable it for all roles, right?

**Comment by kuzmany** [comment]
@escopecz migrations now fixed

**Comment by RCheesley** [comment]
@kuzmany sorry there's a conflict here but we should be able to get it tested today if you can fix it!

**Comment by kuzmany** [comment]
@RCheesley Conflicts resovled

**Review by RCheesley — APPROVED** [review]
Exporting worked for those with permission, and was removed for those without. Thanks for the PR @kuzmany !

**Review by escopecz — APPROVED** [review]
The migration works as well:
Screenshot 2024-04-22 at 16 53 46

**Comment by RCheesley** [comment]
Does this also impact the API? I've written the basic end-user docs update: https://github.com/mautic/user-documentation/pull/295

**Pull request merged**